### DISCUSSION TITLE
Please allow rotation of GpuParticles3D to rotate around all 3 axes

### DISCUSSION THREAD

**Discussion by yosimba2000** [godotengine/godot-proposals#7979] [3D]
Right now, GpuParticles3D can only rotate around their local y-axis for some reason. It would be better to allow more creative freedom by allowing rotation around the X and Z axes as well.

**Comment by Calinou** [comment]
I believe https://github.com/godotengine/godot/pull/79527 allows for this. Can you [test](https://docs.godotengine.org/en/latest/contributing/workflow/testing_pull_requests.html) that PR locally and check?

**Reply by yosimba2000** [reply]
I've just tried it for about 20 minutes, and from the Github description it looks like it should do what I want (orbital velocity), but I can't get it to work. The best I can do is get it to spaz out when CurveXYZ texture is applied to Orbital Velocity.

**Reply by QbieShay** [reply]
I didn't implement per-axis rotation yet. You can rotate particles around an axis with orbit veloity, but i am not sure how it behaves if the particle sits at the origin. It is not meant to be used for that.

Allowing multiple rotation axes than just Y is in my long wishlist of things I'd like to do, but it cannot be done in the particle shader it needs to be done either at the transform pass (where we how have the y to velocity + z billboard option) or in the render shader

**Comment by Robert-K** [comment]
I was very frustrated to find out that's a limitation. A major use case for 3D particles is debris, and it looks completely wrong without true individual angular velocities.

Is there any solution/workaround in the meantime?

**Comment by elvisish** [comment]
Is it even documented that rotating particles z-axis doesn't actually do anything? Making splash particle sprays is impossible to get the rotation correct without being able to freely rotate them as needed.

**Reply by QbieShay** [reply]
I added 3d rotation in the latest dev build. Let me know if it works for you

**Reply by Calinou** [reply]
For reference, this was implemented in https://github.com/godotengine/godot/pull/112447 (see also https://github.com/godotengine/godot/pull/116620).

**Comment by yosimba2000** [comment]
thanks Qbieshay and to everyone else for working on this :)
